src/Cookie/CookieJar.php

```diff
@@ -0,0 +1,283 @@
+<?php
+namespace Clue\React\Buzz\Cookie;
+
+use Psr\Http\Message\RequestInterface;
+use Psr\Http\Message\ResponseInterface;
+use Psr\Http\Message\UriInterface;
+
+/**
+ * CookieJar is responsible for automatic handling HTTP cookies 
+  */
+class CookieJar
+{
+    /** @var CookieSet[] cookies array */
+    private $cookies;
+
+    /** @var string filename where the cookis will be stored */
+    private $filename;
+
+
+    /** @var bool indicates if we need throw exception or just ignore invalid cookies  */
+    private $strictMode = false;
+
+    /** @var bool indicates if we save session cookies in the cookie-file or not  */
+    private $persistSessionCookies = false;
+
+    /**
+     * @param string $cookiesFilename Filename where the cookies are stored.
+     */
+    public function __construct($cookiesFilename = null)
```

Comment:
Filesystem is inherently blocking.

As a first step I would question whether we need filesystem access in the first place It's my understanding that this could be handled entirely in-memory and we can also move persistence logic to a different layer and/or address this later on. What do you think?